File: 200. Number of Islands.py
Problem link: https://leetcode.com/problems/number-of-islands/
Problem Difficulty: Medium

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if grid == None or not len(grid):
            return 0
        ans = 0
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == '1':
                    ans += self.explore(i, j, grid)
        return ans
    def explore(self, i, j, grid):
        if i < 0 or j < 0 or i >= len(grid) or j >= len(grid[i]) or grid[i][j] == '0':
            return 0
        grid[i][j] = '0'
        self.explore(i, j + 1, grid)
        self.explore(i, j - 1, grid)
        self.explore(i + 1, j, grid)
        self.explore(i - 1, j, grid)
        return 1

File: Leetcode/200. Number of Islands.java
/*
 Problem Link: https://leetcode.com/problems/number-of-islands/
 Problem Type: Depth First Search, Recursion
*/

class Solution {
    public int numIslands(char[][] grid) {
        int numberOfIslands = 0;
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == '1') {
                    numberOfIslands++;
                    depthFirstSearch(grid, i, j);
                }
            }
        }
        return numberOfIslands;
    }

    private void depthFirstSearch(char[][] grid, int i, int j) {
        if (i < 0 || j < 0 || i >= grid.length || j >= grid[0].length || grid[i][j] != '1') {
            return;
        }
        grid[i][j] = 'n';
        depthFirstSearch(grid, i + 1, j);
        depthFirstSearch(grid, i - 1, j);
        depthFirstSearch(grid, i, j + 1);
        depthFirstSearch(grid, i, j - 1);
    }
}
